### DISCUSSION TITLE
SSR vs SSG

### DISCUSSION THREAD

**Discussion by JHerwig** [nuxt/nuxt#9577] [Questions (Nuxt 2)]
Hey folks,
For each new project, I have to decide which concept I use. I know every use case can be achieved whatever concept is used. But some are easier than others. This post should help every beginner to choose the best concept for their use case.

I would like to know your recommendations:
- When to use Static-Site-Generation (with Axios for dynamic content) and when to use Server-Side-Rendering?
- When to use an API or build the complete logic inside the service?

Are there performance differences?
Are there security issues?
How changes the deployment?
Are there more advantages and disadvantages?

Let the discussion begin!

**Comment by TheAlexLichter** [comment]
My decision tree is simple:

1. Is the site an internal application or not public facing (everything behind a login)?

**YES?** Then I'd go with an SPA (so no server side rendering at all).
**NO?** Continue with 2.

2. Do I need (near) real-time data that is critical for SEO _or_ is my data changing at quickly (e.g. in minutes, not hours)? Think about sth. like eCommerce or a forum here.

**YES?** Continue.
**NO?** Then you can go full static (SSG) and rebuild on data change. Great examples: Portfolio Page, Company Site, Course Platform (with a new course every month)


For anything else I personally would go with dynamic SSR (generating HTML on the fly) or ISR (SSG + revalidating data).

---

I'd love to here some other opinion though 😋

**Reply by kissu** [reply]
Agreed with the aforementioned approaches.
Btw ISR sounds interesting, how would you re-validate the data?

One use case I thought about lately but if you want to have some content hidden behind an authentication, like private documentation or resources (like videos), don't go static, either SPA or SSR and fetching the content on the fly.

**Reply by TheAlexLichter** [reply]
You'd revalidate the data by re-generating "just this site". Either because the site had never been generated (think of a dynamic site that didn't exist during the initial build) or after a certain time after it had been generated (e.g. generate a new one after 10 minutes, on the next request).

**Reply by JHerwig** [reply]
Your decision tree makes sense. ISR is new to me. Maybe it is worth adding an entry about it inside the official docs?

**Reply by TheAlexLichter** [reply]
ISR is a "kind of" new concept and not that easy to realize with Nuxt at the moment (but it will be easier with Nuxt Nitro). So this will likely be covered as soon as the first class support is present

**Reply by vxncetxn** [reply]
@manniL Does the newly-released Nuxt 3 public beta support the Next.js-style ISG/ISR that you described? I read the new docs site but it doesn't really mention anything about it.

**Reply by TheAlexLichter** [reply]
Works via `routeRules` ~Not yet, still WIP! 😊~
![grafik](https://user-images.githubusercontent.com/640208/139432910-0c569022-b7db-405b-a04b-c1ad619f01d7.png)

**Comment by pablo-vega-kloeckner-i** [comment]
@manniL 
Your decision tree for choosing between SPA, SSG, SSR, or ISR seems logical and well-structured. However, technology and best practices evolve, so it's good to periodically review and validate your approach. Could you please confirm if this decision tree still holds true, or if there have been any updates or changes in best practices that I should be aware of?

On the other hand, Could you kindly explain me why you recommend to go for SPA for internal application or not public facing?

**Reply by kissu** [reply]
Those rendering modes are still totally valid. 👌🏻 

As for internal apps since they're behind a login wall there is no point into speeding the initial rendering of the page or having SEO.

**Reply by TheAlexLichter** [reply]
As @kissu said! Only addition is "hybrid" is now possible via routeRules 🙌

**Reply by kissu** [reply]
https://nitro.unjs.io/config#routerules
and
https://nuxt.com/docs/guide/concepts/rendering#hybrid-rendering

**Reply by pablo-vega-kloeckner-i** [reply]
@manniL  @kissu  Thank you very much for taking the time to answer. I just have one doubt, can you say that CSR for such environments (initial rendering of the page or having) is better in terms of performance?

My doubt comes to that in first instance I was taking SSR as a mechanism to improve the speed in which the pages load and the user interacts with them.  

> Benefits of server-side rendering:
Performance: Users can get immediate access to the page's content because browsers can display static content much faster than JavaScript-generated content. At the same time, Nuxt preserves the interactivity of a web application when the hydration process happens. (Taken from official documentation )

Can it be the case that in SSR the pages load faster but in CSR the interactions are faster?

**Reply by kissu** [reply]
@pablo-vega-kloeckner-i **initial** performance behind an authentication does not really matter. What matters is when your app is running and navigating between pages. So having CSR/an SPA there is totally fine.
If it's for a public-facing website, it's not the best to have an SPA-only. Better to have SSR/SSG + SPA hydrated afterwards so that the **initial** rendering is fast but the page navigations are too (thanks to `vue-router`).

So yes, you have 3 approaches overall:
- SPA-only, wait long initially then go fast for the navigations
- MPA-only, which is the approach used by AstroJS by default where you load each page all over again from a server with old-school browser navigation, takes a bit more time between each navigation afterward
- Nuxt uses an approach of having both: server rendering for the first page to boost the initial speed, then having an SPA for fast navigations once the app is _hydrated_

SSR is mostly to have some good SEO.
Hence why websites should have some kind of SSR.
While apps, could be SPA-only.
Who wants to SSR Gmail or Facebook Messenger?  No point, keep it then as an SPA-only. (you could use SSR for the initial login page but eh, not THAT much of a speed benefit overall  🤷🏻‍♂️)
